hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java

```diff
@@ -878,15 +884,23 @@ private synchronized void flushWrittenBytesToServiceInternal(final long offset,
         if (!successValue) {
           LOG.debug("A past append for the given offset {} with blockId {} " +
                   " for the path {} was not successful", offset, failedBlockId, path);
-          throw new IOException("A past append was not successful");
+          throw new IOException("A past append was not successful for blockId " +
+                  failedBlockId + " and offset " + offset + " for path" + path);
         }
         // Generate the xml with the list of blockId's to generate putBlockList call.
         String blockListXml = generateBlockListXml(blockIdList);
+        TracingContext tracingContextBlobFlush = new TracingContext(tracingContext);
+        tracingContextBlobFlush.setFallbackDFSAppend("B " + position);
         op = client.flush(blockListXml.getBytes(), path,
-                isClose, cachedSasToken.get(), leaseId, getETag(), new TracingContext(tracingContext));
+                isClose, cachedSasToken.get(), leaseId, getETag(), tracingContextBlobFlush);
         setETag(op.getResult().getResponseHeader(HttpHeaderConfigurations.ETAG));
-        getMap().clear();
-        orderedBlockList.clear();
+        try {
+          mapLock.lock();
```

Comment:
why is lock required here?
1. `flushWrittenBytesToServiceInternal` is synchronized. All other important methods like `write` are synchronized.
2. Yes there can be parallel threads that are calling append API. But, before coming to this method, its checked that all the parallel appends are either failed or succeeded:
   1. via flush() method:
        1. `synchronized: flushInternalAsync()` -> **`waitForAppendsToComplete`** ---> `flushWrittenBytesToServiceAsync()` ->  `flushWrittenBytesToServiceInternal()`
   2. via hsync(), hflush(), close():
        1. `flushinternal()` -> `flushWrittenBytesToService()` -> **`waitForAppendsToComplete();`** --> `flushWrittenBytesToServiceInternal`

Reply:
Added just for sanity for all map operations.